include/ginkgo/core/reorder/reordering_base.hpp

```diff
@@ -59,7 +59,14 @@ namespace reorder {
  * It contains a factory to instantiate the reorderings. It is up to each
  * specific reordering to decide what to do with the data that is passed to it.
  */
-class ReorderingBase : public EnableAbstractPolymorphicObject<ReorderingBase> {
+template <typename IndexType = int32>
+class ReorderingBase
+    : public EnableAbstractPolymorphicObject<ReorderingBase<IndexType>> {
+public:
+    using index_type = IndexType;
+
+    virtual const array<index_type>& get_permutation_array() const = 0;
```

Comment:
you can also provide the setter for permutation_array in protected and the permutation data in private.
without implement it in each class